Create a collection of navigation routines  for FLL robotics competition and others

pacojavi opened the pull request:
The objective is to define navigation functions, with or without POSE, for robotics competitions like FLL, 
The basis maybe: Initialize, Straight Travel with Gyro Sensor, Precision Turn, Line Follower,  LAG Travel, Menu for missions, and Aux motor functions.

pacojavi committed:
Add robot.h, robot.c i

Initial Deploy of robot for FLL

Co-Authored-By: RedTeam FLL <68497796+redteamsp@users.noreply.github.com>

pacojavi committed:
Corrected errors in robot.h

Doxygen errors found and corrected

Co-Authored-By: RedTeam FLL <68497796+redteamsp@users.noreply.github.com>

JakubVanek commented on include/ev3_robot.h:
```diff
@@ -0,0 +1,95 @@
+ * @param Distance Distance to travel in mm 
+ * @return distance to travel in degrees.
+ */
+int Travelcm(int Distance, uint8_t angle, int power);
```
I feel that the current name `Travelcm` hints that this function does start the movement, but that is not the case. `int CalculateTravelDegrees(int mm)` would feel better to me here. Also, the function in `ev3_robot.c` is defined without the last two arguments, causing the Travis build to fail.

pacojavi replied:
corrected! You are right CalculateTravelDegrees is more suitable

JakubVanek commented on include/ev3_robot.h:
```diff
@@ -0,0 +1,95 @@
+void RobotInit(int ColorLeft, int ColorRight, uint8_t MotorLeft, uint8_t MotorRight, int Gyro, bool Debug);
+
+/**
+ * @brief Initialize PID for navigation with girosensor.
```
Only a typo here, `gyrosensor`

pacojavi replied:
This function has been changed, now dont use parameters. All necessary robot data can be found in my_robot.h.
So, that's a definition file, do you think is better to name it like ev3_my_robot.h or my_robot.h, is correct?

JakubVanek replied:
There is a problem with this approach, see https://github.com/c4ev3/EV3-API/pull/53#discussion_r473805719. That being said, the definition macros could be instead replaced with a struct that would be passed to the initialization function. The struct would be defined in `ev3_robot.h`:
```c
typedef struct {
    float WheelDiameter;
    float Width;
    int LeftMotorPort;
    int RightMotorPort;
    ...
} ROBOT_PARAMS;
```

In the user code, the structure would be used to define a new variable containing the parameters:
```c
ROBOT_PARAMS my_robot = {
    .WheelDiameter  = 56.0f, // mm
    .Width          = 100.0f, // mm
    .LeftMotorPort  = OUT_A,
    .RightMotorPort = OUT_B,
     ...
};
```

The initialization function then could accept a pointer to that structure:
```c
void RobotInit(ROBOT_PARAMS *params) {
    GlobalVars.WheelDiameter = params->WheelDiameter;
    ...
}
```

And the initialization function would be called like this:
```c
    // initialize with my robot definition
    RobotInit(&my_robot);
```

pacojavi replied:
GlobalVars is a variable of type ROBOT_PARAMS, correct? and in the rest of functions in robot c, we will use this global variable, right?
Well, we agreed with that, for the next commit, so this wil require an example for use, in which directory we can place it?

JakubVanek replied:
> GlobalVars is a variable of type ROBOT_PARAMS, correct?

It can be, but it doesn't have to. Originally I thought of `ROBOT_PARAMS` as containing only unchanging data such as the wheel diameter. `GlobalVars` would contain the changing state (pose, ...) and then either a pointer to `ROBOT_PARAMS` or a copy of the parameters contained within.

> and in the rest of functions in robot c, we will use this global variable, right?

It is one of the possible solutions. Currently `MyRobot` behaves exactly like the `GlobalVars` would behave.

> Well, we agreed with that, for the next commit, so this wil require an example for use, in which directory we can place it?

`/examples/ev3_robot.c` should do.

JakubVanek commented on include/ev3_robot.h:
```diff
@@ -0,0 +1,95 @@
+ * @param Kd Derivative part constant, default value 0 
+ * @return N/A.
+ */
+void SetKPID(short Kp, short Ki,short Kd);
```
I would rename this function too (`void Set<or Configure>StraightPID(int Kp, int Ki, int Kd)`), as having multiple PID controllers could be useful. One of them is the straight travel PID as implemented here, but turning with the help of the gyro sensor may require a different PID.

pacojavi replied:
That is the idea, multiple PID can be used, for Gyro Sensor, Color Sensor on line follower, ultrasonic on wall follower...
Also corrected the type of parameters, they must be float, not short.

JakubVanek commented on include/ev3_robot.h:
```diff
@@ -0,0 +1,95 @@
+ * @param Debug If True print on the LCD screen the result of initialiation
+ * @return N/A.
+ */
+void RobotInit(int ColorLeft, int ColorRight, uint8_t MotorLeft, uint8_t MotorRight, int Gyro, bool Debug);
```
I wouldn't integrate `ColorLeft` and `ColorLight` (unless optional with `-1`) into this interface. I think it depending on them would decrease the usefulness of this module if the robot has no color sensors/lines to track.

pacojavi replied:
Well, good point, our base robot has 2 sensor colors and a gyro sensor, but in other configurations maybe three sensor colors, or two gyro sensors or ultrasonic. 
In the robot definition file, we can put some options and then check it.
In the output configuration, we have a doubt, from programing point of view, is there a difference between medium and large motors? we dont found any reference

JakubVanek replied:
At this level there indeed should be no difference. There is a difference at the kernel driver level - medium motor has slightly different PID parameters for its speed PID controller (it internally converts the OnFwdReg speed request to the actual PWM duty cycle).

Ideally one would call `OutputSetTypes()` with the correct type (`TYPE_TACHO` (7) for large motor, `TYPE_MINITACHO` (8) for medium motor) to tell the motor kernel driver what motors are plugged in. In practise, the correct configuration is persisted from the LMS2012 menu process, therefore it should not be necessary to call `OutputSetTypes()` unless you want to hot-swap the motors when the C4EV3 program is running.

JakubVanek commented on include/ev3_robot.h:
```diff
@@ -0,0 +1,95 @@
+/**
+ * @brief Initialize all robot variables, set POSE to (0,0,0),  * reset gyrosensor and reset tacho counters.
+ * @param port Port to which the sensor is connected
+ * @param mode How to measure the rate
```
These two doc comments refer to parameters that do not exist anymore.

JakubVanek replied:
resolved now

pacojavi replied:
Cut&Paste! corrected.

JakubVanek commented on API/ev3_robot.c:
```diff
@@ -0,0 +1,163 @@
+	PoseInit();
+	SetKPID(0,0,0);
+
+	MyRobot.WheelCircunference = MyRobot.WheelDiameter * M_PI;
```
`WheelDiameter` is not set here yet (unless MyRobot is used in user code too)

JakubVanek replied:
resolved now

JakubVanek commented on API/ev3_robot.c:
```diff
@@ -0,0 +1,163 @@
+	
+	ResetEV3GyroSensor(MyRobot.Gyro, EV3GyroHardwareCommand0x88); // chack Gyro Veersion //
+	ResetEV3GyroSensor(MyRobot.Gyro, EV3GyroHardwareReboot); // 2.5s, valid for all versions
+	ResetEV3GyroSensor(MyRobot.Gyro, EV3GyroSoftwareOffset); /* chack for proper reset */
```
The software reset here is correct, the hardware reboot above does not reset the software offset.

JakubVanek replied:
I think that gyro calibration should be done in a separate user-callable function. The reason I see for this is to allow the program to have a separate `Press OK to start calibration` prompt. Before proceeding, users could ensure that no external movement would disturb the calibration process.

pacojavi replied:
We have doubt, about this. In EV3-G we have a block to make the calibration and and entry menu to do it. 
Always previous at first mission in FLL competition, we make a Gyro Reset. It is a headhache when the gyro sensor goes crazy, and there's no way of knowing when it's going to go faulty
Thanks for your explanation!

JakubVanek replied:
Indeed, the true recalibration should theoretically be done only once at the program start, subsequent resets should be `EV3GyroSoftwareOffset`. In practise, it may be the case that the angle does start to drift away after a while. In that case, either a new recalibration is needed, or one needs to do the rotation rate -> angle integration themselves and do own rate offset detection (idle sensor wouldn't report 0 °/s).

pacojavi replied:
we tried a experiment in EV3-G, and the gyro sensor when robot moves is not equal to 0 deg/s, it has a lot of noise, it may be that we are doing something wrong. The problem is where the sensor is drifting, in the middle of a straight move with gyrosensor. Maybe a solution is read the rotation rate and if it is over a value, reset the gyro sensor, but unfortunately is impossible to return to the 0 position, because you must stop the motors to reset the gyro, and the gyro is drifting in the meanwhile. We try that there are no sudden movements that seem more problematic.
We take this note, if it seems necesary to make a reset in other part of the program  we will use EV3GyroSoftwareOffset,

JakubVanek replied:
>The problem is where the sensor is drifting, in the middle of a straight move with gyrosensor

If you stop the robot, does the rate return to 0 deg/s? If yes, then the problem might on the physics side of things (see hypothesis below), but if the rate is still nonzero, then the gyro is the problem. Also, is the drift happening all the time during the robot travel, or is there a single time point at which it starts drifting?

This looks like something I have observed on the leJOS experiment mentioned previously. IIRC my robot was lightly steering to the right instead of going straight.

I have a hypothesis about why this is happening, but I haven't analyzed or tested it. The core of the problem may be that the lego gyro behaves like a single-axis gyroscope (ignoring the TILT-ANG/TILT-RATE modes). This has interesting consequences - one can turn the gyro three times along its non-sensing axes such that the gyro is rotated 90° to the original position; yet the reported sensor angle should stay unchanged. If something similar on much smaller scale happens due to vibration generated when the robot is moving, the accumulated angle may get slowly skewed towards one direction.

JakubVanek commented on API/ev3_robot.c:
```diff
@@ -0,0 +1,163 @@
+
+}
+
+int StraightGyroDegrees(int distDegre, int angle, int speed, bool brake){
```
This is an interesting idea. However, does it work in practise? I tried to do something similar on leJOS a while ago and I think that the robot was riding less straight with the gyro control than when riding without it. It is however entirely possible that I did something wrong (tuning, calibration, etc.) that made it less than ideal for me. Also, if your robot can hit some movable obstacles that would change the direction of the robot, then this seems like a valid approach.

pacojavi replied:
We use this in the past year, only in proportional mode, the distances on the mat, dont need integral or derivative. And it goes well, with a Kp of 2, and the EV3 move steering module. With Kp over 5 the robot swing.
Our function has a ramp acceleration, proportional travel and ramp deceleration. 
We are looking for similar code to move steering in output.c but we cannot found it. 
Another idea is to use line follow and gyro at same time. Go straight is not easy in Lego.

JakubVanek replied:
Oki, let's then go ahead with this!

JakubVanek commented on API/ev3_robot.c:
```diff
@@ -0,0 +1,163 @@
+	
+	error = angle - angleNow; // For Proportional part
+	errorD = error - oldErrorD; // For Darivative part
+	errorI = errorI + error; //For Integral part
```
I think adding https://en.wikipedia.org/wiki/Integral_windup protection could help when using the integral term. I think I had good experience with one of these:

* not adding to the integral when motors are activated beyond some power threshold
* not allowing the integral itself to go over some threshold in its absolute value

pacojavi replied:
Yes, we agree with that. This a very first version, we need to check it a more and more

JakubVanek commented on API/ev3_robot.c:
```diff
@@ -0,0 +1,163 @@
+	// new power to motors
+	// need to cheek max/min speed?
+	OnFwdReg(MyRobot.MotorLeft, speed + uOut);
+	OnFwdReg(MyRobot.MotorRight, speed - uOut);
```
Beware, without adding some additional speed ramping, your robot is going to overshoot the position badly at higher speeds. What I mean by ramping is gradually decreasing the baseline speed when the robot is getting near the target position. Ramping at the beginning (gradually increasing the speed) should also help on slippery surfaces.

There is a under-documented function `RotateMotorRampNoWait()` that instructs the EV3 kernel driver to do part of the ramping for you. However, doing it here manually should also be OK.

pacojavi replied:
Our strategy at EV3-G was to increase speed by 45 degrees and decrease in similar way, this was precise, but it is slow, every turn, every change of direction, the robot stops before.
We will try to code  our EV3-G block.
Many, many thanks for the suggestion of function RotateMotorRampNoWait()
we are still looking for documentation about ev3_output.

JakubVanek replied:
If you can avoid braking altogether (indeed, the current version allows this), that would be the fastest. For the stopping before the target, it may be necessary to have some minimum continuous speed level at the end of the deceleration ramp. It's true that it won't make the whole process faster though.

Re documentation: it's quite low-level, but the kernel stuff that `ev3_output.c` in the end calls is here: https://github.com/mindboards/ev3sources-xtended/blob/b32a23625be02eb22f23ac45d2ef3bd4a2a9173f/ev3sources/lms2012/d_pwm/Linuxmod_AM1808/d_pwm.c#L2146 There should be some similarities to EV3-G blocks because they have to go through the kernel too.

pacojavi replied:
Thanks a buch!!!
A last question about output.c, in MS VSCODE show me the error:
{
	"resource": "/c:/GitHub/EV3-API/API/ev3_output.c",
	"owner": "C/C++",
	"severity": 8,
	"message": "the identifier \"O_SYNC\" is not defined",
	"startLineNumber": 140,
	"startColumn": 66,
	"endLineNumber": 140,
	"endColumn": 72
}

JakubVanek replied:
I think you can remove the problematic `O_SYNC`, it should have no effect there. I don't know why the warning appeared, but it might be that the include path set used inside vscode isn't the same as the one the compiler uses.

pacojavi replied:
Agree, maybe a problem with vscode, we check all the files who uses O_SYNC and all have this warning, its curious that O_RDWR does not have this issue

pacojavi replied:
If the PID converges then there should be no speed problems, we adjust the K's by try & error, anyway we would like the steps in regulation will be more soft. The minimun is 2 speed differences between motors ( +1, -1)

JakubVanek commented on API/ev3_robot.c:
```diff
@@ -0,0 +1,163 @@
+
+int uOut; // control for motors
+
+OnFwdSync(MyRobot.MotorDual, speed);
```
I would use `OnFwdReg` here too. `OnFwdSync` does inter-wheel synchronization at the kernel level. However, I think this is disabled after the first loop ends (due to the call to `OnFwdReg`). As the gyro-correction does sort-of inter-wheel synchronizatino as well, I would rely on `OnFwdReg` only.

pacojavi replied:
Ok, We were trying to figure out how the motor functions work. In other languages, implementations is not possible to go sync and then control each motor individually.

JakubVanek replied:
> is not possible to go sync and then control each motor individually

This should be implementable (move tank in EV3-G likely does something like this) by using `OnFwdSyncEx(ports, speed, turn_percent, reset)`.

JakubVanek replied:
I have realized there may be some issues with calling `OnFwdSyncEx` in a loop, it would be necessary to test this.

JakubVanek commented on API/ev3_robot.c:
```diff
@@ -0,0 +1,163 @@
+	ROBOT_POSE Pose;
+	int Width;
+	float WheelDiameter;
+	float WheelCircunference;
```
only a typo here, `WheelCircumference`

pacojavi replied:
Solved!

JakubVanek reviewed (commented):
Hi @pacojavi,

welcome and thank you for contributing to C4EV3! I have shared some of my thoughts in the comments.

Best regards,

Jakub

pacojavi committed:
Redefine init functions and added new header file my_robot.h

The my_robot.h contains all the parameters of the robot, ports used, size, and PID constant.
The init functions now use this file often parameters

Co-Authored-By: RedTeam FLL <68497796+redteamsp@users.noreply.github.com>

JakubVanek commented on include/my_robot.h:
```diff
@@ -0,0 +1,74 @@
+/** \file my_robot.h
```
I'm afraid that this will not work for [Eclipse plugin](https://github.com/c4ev3/EV3-Eclipse-Plugin) users (it works for the Makefile project though). I'm not 100% sure, but it seems that the plugin ships precompiled `libev3api.a` and it does not recompile EV3-API from the source code. This would mean that the values in `ev3_robot.c` would always get initialized to the default values.

pacojavi replied:
my_robot.h is not used anymore! Init functions redefined, and split mechanical parameters from internal parameters.

pacojavi committed:
Major changes and corrections

dont use anymore my_robot.h instead a new ROBOT_PARAMS type def must be used. See Example ev3_robot.c
new initialization procedural
See Example ev3_robot.c
Add some sensors use cases, and the arm motors ports included.
Corrections in some functions
Added at first version of example of use

Co-Authored-By: RedTeam FLL <68497796+redteamsp@users.noreply.github.com>

JakubVanek commented on API/ev3_robot.c:
```diff
@@ -51,89 +57,122 @@ typedef struct {
-OnFwdSync(MyRobot.MotorDual, speed);
+/* OnFwdSync(Robot.MotorDual, speed);*/
+OnFwdReg(Robot.MotorLeft, speed);
+OnFwdReg(Robot.MotorRight, speed);
```
I think you should be able to use `Robot.MotorDual` here too.

pacojavi replied:
We will try! thks.

pacojavi committed:
Exampe Tested And Straight bugs corrected

Corrected
- mm to Degrees translation
- Aplication of control signal in StraightbyGyroDegrees is now correct
robot library sucesfully compiled
example sucesfully compiled

Co-Authored-By: RedTeam FLL <68497796+redteamsp@users.noreply.github.com>

a3f commented:
@JakubVanek Thanks very much for the review! @pacojavi please ping me when you have finalized this,

pacojavi committed:
Added Turning functions and straight, line follower and LAGS

Added Spin and pivot robot rotation funcionts, new straightbygyro, line follower and LAGS, and some useful functions

pacojavi committed:
Warnings Fixed

pacojavi committed:
Added Menu Button Style for several runs.

pacojavi committed:
Added params to use  driving gears

pacojavi committed:
Update ev3_robot.c

minor changes and added gears to power train motor

pacojavi committed:
Fix minor bugs

Bugs in fucntions declarations, and some imporvementes in function to travel straigh until a line is reached

Co-Authored-By: RedTeam FLL <68497796+RedTeamSP@users.noreply.github.com>

pacojavi committed:
Release 1.0.0

First release, see in examples directory for use.
ToDo: iimplement POSE and maybe some optimization

Co-Authored-By: RedTeam FLL <68497796+RedTeamSP@users.noreply.github.com>
Co-Authored-By: Carmen <65446318+Carmoruda@users.noreply.github.com>
Co-Authored-By: MichaelVDC <65283699+MichaelVDC@users.noreply.github.com>
Co-Authored-By: MiguelRedteam <73019235+MiguelRedteam@users.noreply.github.com>

pacojavi committed:
Impporvementes in linefollowers ans LAGS functions.
Also color sensor calibration is provided

pacojavi committed:
Several improvements in linefollowers and LAGS, also a routines to calibrate dolor sensors is provided

pacojavi committed:
Calibration values saved to a file

use a file called lightcalibration.cal to store ans retrieve the values of raw min and raw  max values of sensors
The file is located in /home/root/lms2012/prjs.

pacojavi committed:
Añadido inicializacion Color y nueva salida final

pacojavi committed:
Se añade uso de sensor color calibrado, y nueva version de salida final

pacojavi committed:
Update main.c

Corregido error en brazo neumatico

pacojavi committed:
a ignorar

pacojavi committed:
Revisada Salida Final - MV

pacojavi committed:
Correccion Salida MG

pacojavi committed:
Correccion Salida MG

pacojavi committed:
Added functions for Aux Motors

pacojavi committed:
Format chamges

pacojavi committed:
Salida Final pre Release

pacojavi committed:
Update main.c

Salida 1 hasta prueba de arrastrar rueda

pacojavi committed:
Sallida Carmen

Revision salida Carmen

pacojavi committed:
Update Salida Carmen

Salida Carmen lista para tablero

pacojavi committed:
prueba 3

pacojavi committed:
Changes in FollowLine Algorithm

pacojavi committed:
Version Beta Toledo

Pruebas para la FLL CLM

pacojavi committed:
Update main.c

pacojavi committed:
Minor bugs in doc

Corrected some bugs in doc for doxygen

pacojavi committed:
Version 0.9a Gran Final

pacojavi committed:
version GT 0.9

verison inicial GranFinal

pacojavi committed:
New TurnGyroFast added & new functions for GF

RT21GF update for Tobogan & Carmen
Ev3_robot + TurnGyroFast

pacojavi committed:
New TurnGyro functions added

TurnGyrowithClearance and TurnGyrowithClaranceTimer

pacojavi committed:
Corrections in Turn functions

Some corrections on new turn functions

pacojavi committed:
Update Tobogan function for FLL_GF

Update tobagan exir

JakubVanek commented:
Hi,

I'm sorry for not engaging more, I've been quite busy during the semester, although there was also a bit of procrastination involved :sweat_smile: 

Would you prefer further extending the ev3_robot API, or would it be better to review & merge the API as-is?

With regards,

Jakub

pacojavi commented:
Hi,
Good news to be back in touch.
Yesterday we finished the FLL in Spain, with a very good result, we were champions in our regional tournament, and a very worthy role in the national tournament. A leap forward, mostly thanks to the use of this development.
We think that as a first version it is ready to merge, but there is a lot of work ahead, reordering some things and compacting the code.
We have added some routine to calibrate and save the calibration of the colour sensors, independently, which I think should go in the colour sensors library. It is relative to the reflected light measurement, each sensor has its own range, and the default calibration is common for all the sensors used. What we do is to read the RAW value and scale independently for each sensor. We have a partial solution, but it is not definitive, it is not scalable in the implementation.
But it can be started in a new request.
From now on, for a while, I will work alone on this part without the rest of the team.
Thank you very much for everything
Regards
Javier

a3f commented:
@pacojavi Confgratulations on your placement :) I wasn't involved much, but I am wondering what's the status here? @JakubVanek Did you come around to take a look?
